hls-plugin-api/src/Ide/Plugin/Resolve.hs

```diff
@@ -29,44 +31,61 @@ import           Language.LSP.Server           (LspM, LspT,
-                r@(InR Null) -> pure r
-                (InL ls) | -- If the client supports resolve, we will wrap the resolve data in a owned
-                           -- resolve data type to allow the server to know who to send the resolve request to
-                           supportsCodeActionResolve caps -> pure $ InL ls
-                           --This is the actual part where we call resolveCodeAction which fills in the edit data for the client
-                         | otherwise -> InL <$> traverse (resolveCodeAction (params ^. L.textDocument . L.uri) ideState pid) ls
-    in (mkPluginHandler SMethod_TextDocumentCodeAction newCodeActionMethod
-    <> mkResolveHandler SMethod_CodeActionResolve codeResolveMethod)
-    where
-        dropData :: CodeAction -> CodeAction
+mkCodeActionHandlerWithResolve recorder codeActionMethod codeResolveMethod =
+  let newCodeActionMethod ideState pid params = runExceptT $
+        do codeActionReturn <- ExceptT $ codeActionMethod ideState pid params
+           caps <- lift getClientCapabilities
+           case codeActionReturn of
+             r@(InR Null) -> pure r
+             (InL ls) | -- If the client supports resolve, we will wrap the resolve data in a owned
+                        -- resolve data type to allow the server to know who to send the resolve request to
+                        supportsCodeActionResolve caps -> pure $ InL ls
+                        --This is the actual part where we call resolveCodeAction which fills in the edit data for the client
+                      | otherwise -> InL <$> traverse (resolveCodeAction (params ^. L.textDocument . L.uri) ideState pid) ls
+  in (mkPluginHandler SMethod_TextDocumentCodeAction newCodeActionMethod
+  <> mkResolveHandler SMethod_CodeActionResolve codeResolveMethod)
+  where dropData :: CodeAction -> CodeAction
         dropData ca = ca & L.data_ .~ Nothing
         resolveCodeAction :: Uri -> ideState -> PluginId -> (Command |? CodeAction) -> ExceptT ResponseError (LspT Config IO) (Command |? CodeAction)
         resolveCodeAction _uri _ideState _plId c@(InL _) = pure c
         resolveCodeAction uri ideState pid (InR codeAction@CodeAction{_data_=Just value}) = do
-            case A.fromJSON value of
-              A.Error err -> throwE $ parseError (Just value) (T.pack err)
-              A.Success innerValueDecoded -> do
-                resolveResult <- ExceptT $ codeResolveMethod ideState pid codeAction uri innerValueDecoded
-                case resolveResult of
-                  CodeAction {_edit = Just _ } -> do
-                      pure $ InR $ dropData resolveResult
-                  _ -> throwE $ invalidParamsError "Returned CodeAction has no data field"
-        resolveCodeAction _ _ _ (InR CodeAction{_data_=Nothing}) = throwE $ invalidParamsError "CodeAction has no data field"
+          case A.fromJSON value of
+            A.Error err -> logAndThrow recorder LogParseError $ parseError (Just value) (T.pack err)
```

Comment:
I'm unsure if this is the best place to log. _Hopefully_ caught exceptions should be logged in general, so I think it's probably overkill to always log before throwing, then I think we'll get things twice?

Reply:
Right